config/diff.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type ConfigDiffs []ConfigDiff
+
+type ConfigDiff struct {
+	Path      string      `json:"path"`
+	BaseVal   interface{} `json:"base_val"`
+	ActualVal interface{} `json:"actual_val"`
+}
+
+func diff(base, actual reflect.Value, label string) ([]ConfigDiff, error) {
+	var diffs []ConfigDiff
+
+	if base.IsZero() && actual.IsZero() {
+		return diffs, nil
+	}
+
+	if base.IsZero() || actual.IsZero() {
+		return append(diffs, ConfigDiff{
+			Path:      label,
+			BaseVal:   base.Interface(),
+			ActualVal: actual.Interface(),
+		}), nil
+	}
+
+	baseType := base.Type()
+	actualType := actual.Type()
+
+	if baseType.Kind() == reflect.Ptr {
+		base = reflect.Indirect(base)
+		actual = reflect.Indirect(actual)
+		baseType = base.Type()
+		actualType = actual.Type()
+	}
+
+	if baseType != actualType {
+		return nil, fmt.Errorf("not same type %s %s", baseType, actualType)
+	}
+
+	switch baseType.Kind() {
+	case reflect.Struct:
+		if base.NumField() != actual.NumField() {
```

Comment:
If a new field is added to config struct, then the config is updated post-server-upgrade, will the diff fail because the structs don't match?

Reply:
I think that check is redundant since we deal with the same `model.Config` object  but I left it just to be extra safe.
If something in the serialized config doesn't match with the model then it will just get discarded during deserialization.

Reply:
nit: Shall we do this like below? Because we are using switch against `baseType.Kind()`
```suggestion
		if baseType.NumField() != actualType.NumField() {
```

Reply:
That's a good observation. I am wondering if it makes any difference though. Is the type just an interface that the value implements? If so both ways are semantically equivalent.